### PR TITLE
Enable client credentials grant type

### PR DESCRIPTION
**Pull request by themsaid**
This grant type is useful to allow clients to access data that's not owned by a certain user, for example: searching tweets, listing supported countries, etc...

The PR implements a middleware that can be used to check for client credentials.

### CODE REVIEW

File: src/Http/Middleware/CheckClientCredentials.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        $this->server = $server;
+    }
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request $request
+     * @param  \Closure $next
+     * @return mixed
```
Comment (lucasmichot) [inline src/Http/Middleware/CheckClientCredentials.php]: Missing `@throws \Illuminate\Auth\AuthenticationException`

Comment (themsaid) [inline src/Http/Middleware/CheckClientCredentials.php]: Updated, thanks!

**Comment by taylorotwell** [comment]
Why define a whole new middleware? The TokenGuard will already call the validateAuthenticatedRequest method on the resource server.

**Comment by themsaid** [comment]
For this type of grants there's no User, the TokenGuard expects to be able to successfully retrieve a user.

**Comment by ganey** [comment]
I'm also interested in this, I was using lucadegasperi/oauth2-server-laravel with client_credential tokens before upgrading to 5.3

File: src/Http/Middleware/CheckClientCredentials.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     *
+     * @throws \Illuminate\Auth\AuthenticationException
+     */
+    public function handle($request, Closure $next)
```
Comment (ganey) [inline src/Http/Middleware/CheckClientCredentials.php]: If you wanted to check scopes too, something along the following lines works:
This then works in the same way as the CheckForAnyScope middleware.

```
public function handle($request, Closure $next, ...$scopes)
  {
    $psr = (new DiactorosFactory)->createRequest($request);

    try {
      /**
       * @var $serverRequest ServerRequest
       */
      $serverRequest = $this->server->validateAuthenticatedRequest($psr);
    } catch (OAuthServerException $e) {
      throw new AuthenticationException;
    }

    foreach ($scopes as $scope) {
      if (!in_array($scope,$serverRequest->getAttribute('oauth_scopes'))) {
        throw new AuthenticationException;
      }
    }

    return $next($request);
  }
```

**Comment by enx1998** [comment]
Hi everyone, 
I don't know if this is the right place di place this question, but the argument is the client_credentials grant type.
If I understand well, the client_credential grant was integrated in the code, but I don't know how can I use it to protect my routes.
All I want to do is to make a route available to client that have client_credential token.

Thanks ENx

File: src/Http/Middleware/CheckClientCredentials.php (hunk 2 above)
Comment (endyjasmi) [inline src/Http/Middleware/CheckClientCredentials.php]: This works except when you try to request access token with '*' as scope. Seems like the logic for handling all scope in client credential grant is left out.

**Comment by ankurk91** [comment]
An access_token that was generated by `password` grant can used by bypass this middleware.
Does specs allow this?

**Comment by driesvints** [comment]
@ankurk91 yeah, you can use the same client for multiple grants but it's not recommended. It's best to only use one grant per client.

**Comment by ankurk91** [comment]
Thanks for the answer.

**Pull request merged**